File: demo/test_fizzbuzz.py
"""Tests for fizzbuzz.py — uses stdlib unittest (no external dependencies).

Run with:
    python3 -m unittest test_fizzbuzz.py -v
"""

import unittest

from fizzbuzz import fizzbuzz


class TestFizzBuzzSequence(unittest.TestCase):
    """Verify correct output values at key positions."""

    def setUp(self):
        self.seq15 = fizzbuzz(15)
        self.seq20 = fizzbuzz(20)

    def test_plain_number(self):
        """Numbers not divisible by 3 or 5 should appear as strings."""
        self.assertEqual(self.seq15[0], "1")   # 1
        self.assertEqual(self.seq15[1], "2")   # 2
        self.assertEqual(self.seq15[3], "4")   # 4

    def test_fizz(self):
        """Multiples of 3 (but not 15) should be 'Fizz'."""
        self.assertEqual(self.seq15[2], "Fizz")   # 3
        self.assertEqual(self.seq15[5], "Fizz")   # 6
        self.assertEqual(self.seq15[8], "Fizz")   # 9
        self.assertEqual(self.seq15[11], "Fizz")  # 12

    def test_buzz(self):
        """Multiples of 5 (but not 15) should be 'Buzz'."""
        self.assertEqual(self.seq15[4], "Buzz")   # 5
        self.assertEqual(self.seq15[9], "Buzz")   # 10
        self.assertEqual(self.seq20[19], "Buzz")  # 20

    def test_fizzbuzz(self):
        """Multiples of 15 should be 'FizzBuzz' (the previously buggy case)."""
        self.assertEqual(self.seq15[14], "FizzBuzz")  # 15
        seq30 = fizzbuzz(30)
        self.assertEqual(seq30[29], "FizzBuzz")       # 30

    def test_sequence_length(self):
        """Returned list should contain exactly n elements."""
        for n in [1, 5, 15, 20, 100]:
            with self.subTest(n=n):
                self.assertEqual(len(fizzbuzz(n)), n)


class TestFizzBuzzEdgeCases(unittest.TestCase):
    """Boundary and single-element cases."""

    def test_n_equals_1(self):
        self.assertEqual(fizzbuzz(1), ["1"])

    def test_n_equals_3(self):
        self.assertEqual(fizzbuzz(3), ["1", "2", "Fizz"])

    def test_n_equals_5(self):
        self.assertEqual(fizzbuzz(5), ["1", "2", "Fizz", "4", "Buzz"])

    def test_n_equals_15(self):
        expected_tail = ["Buzz", "11", "Fizz", "13", "14", "FizzBuzz"]
        self.assertEqual(fizzbuzz(15)[9:], expected_tail)


class TestFizzBuzzInputValidation(unittest.TestCase):
    """fizzbuzz() should raise on bad input."""

    def test_raises_type_error_for_float(self):
        with self.assertRaises(TypeError):
            fizzbuzz(5.0)

    def test_raises_type_error_for_string(self):
        with self.assertRaises(TypeError):
            fizzbuzz("10")

    def test_raises_value_error_for_zero(self):
        with self.assertRaises(ValueError):
            fizzbuzz(0)

    def test_raises_value_error_for_negative(self):
        with self.assertRaises(ValueError):
            fizzbuzz(-1)


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: typescript_demo/test_fizzbuzz.ts
/**
 * Tests for fizzbuzz.ts — uses Node's built-in test runner (no external deps).
 *
 * Run with:
 *     npx tsx --test test_fizzbuzz.ts
 *     # or just: npx tsx test_fizzbuzz.ts
 */
import { describe, it } from "node:test";
import assert from "node:assert/strict";

import { fizzbuzz } from "./fizzbuzz.js";

describe("FizzBuzz sequence", () => {
  const seq15 = fizzbuzz(15);
  const seq20 = fizzbuzz(20);

  it("plain numbers appear as strings", () => {
    assert.equal(seq15[0], "1"); // 1
    assert.equal(seq15[1], "2"); // 2
    assert.equal(seq15[3], "4"); // 4
  });

  it("multiples of 3 (but not 15) are 'Fizz'", () => {
    assert.equal(seq15[2], "Fizz"); // 3
    assert.equal(seq15[5], "Fizz"); // 6
    assert.equal(seq15[8], "Fizz"); // 9
    assert.equal(seq15[11], "Fizz"); // 12
  });

  it("multiples of 5 (but not 15) are 'Buzz'", () => {
    assert.equal(seq15[4], "Buzz"); // 5
    assert.equal(seq15[9], "Buzz"); // 10
    assert.equal(seq20[19], "Buzz"); // 20
  });

  it("multiples of 15 are 'FizzBuzz' (the previously buggy case)", () => {
    assert.equal(seq15[14], "FizzBuzz"); // 15
    const seq30 = fizzbuzz(30);
    assert.equal(seq30[29], "FizzBuzz"); // 30
  });

  it("returns exactly n elements", () => {
    for (const n of [1, 5, 15, 20, 100]) {
      assert.equal(fizzbuzz(n).length, n);
    }
  });
});

describe("FizzBuzz edge cases", () => {
  it("n = 1", () => assert.deepEqual(fizzbuzz(1), ["1"]));
  it("n = 3", () => assert.deepEqual(fizzbuzz(3), ["1", "2", "Fizz"]));
  it("n = 5", () => assert.deepEqual(fizzbuzz(5), ["1", "2", "Fizz", "4", "Buzz"]));
  it("n = 15 tail", () => {
    const expectedTail = ["Buzz", "11", "Fizz", "13", "14", "FizzBuzz"];
    assert.deepEqual(fizzbuzz(15).slice(9), expectedTail);
  });
});

describe("FizzBuzz input validation", () => {
  it("throws TypeError for a non-integer number", () => {
    assert.throws(() => fizzbuzz(5.5), TypeError);
  });
  it("throws TypeError for a string", () => {
    assert.throws(() => fizzbuzz("10" as unknown as number), TypeError);
  });
  it("throws RangeError for zero", () => {
    assert.throws(() => fizzbuzz(0), RangeError);
  });
  it("throws RangeError for a negative number", () => {
    assert.throws(() => fizzbuzz(-1), RangeError);
  });
});
